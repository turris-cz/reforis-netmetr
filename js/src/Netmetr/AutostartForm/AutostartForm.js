/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { CheckBox } from "foris";

import AutostartHoursForm from "./AutostratHoursForm";

export default function AutostartForm({
    formData, formErrors, setFormValue, disabled,
}) {
    return (
        <>
            <h3>{_("Autostart")}</h3>
            <CheckBox
                label={_("Enabled")}
                checked={formData.autostart_enabled}
                onChange={setFormValue((value) => ({ autostart_enabled: { $set: value } }))}
            />
            {formData.autostart_enabled
                && (
                    <AutostartHoursForm
                        formData={formData}
                        setFormValue={setFormValue}
                        disabled={disabled}
                    />
                )}
        </>
    );
}
