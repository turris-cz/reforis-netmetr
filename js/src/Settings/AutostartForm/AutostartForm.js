/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { CheckBox } from "foris";
import PropTypes from "prop-types";

import AutostartHoursForm from "./AutostratHoursForm";

AutostartForm.propTypes = {
    formData: PropTypes.shape({
        autostart_enabled: PropTypes.bool,
        hours_to_run: PropTypes.arrayOf(PropTypes.number),
    }).isRequired,
    setFormValue: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

AutostartForm.defaultProps = {
    formData: {
        autostart_enabled: false,
        hours_to_run: [],
    },
    setFormValue: () => {},
};

export default function AutostartForm({ formData, setFormValue, disabled }) {
    return (
        <>
            <h3>{_("Autostart")}</h3>
            <p>
                {_(
                    "Enable Autostart to allow automatic measurement on specific daily hours that can be chosen.",
                )}
            </p>
            <CheckBox
                label={_("Enabled")}
                checked={formData.autostart_enabled}
                onChange={setFormValue((value) => ({
                    autostart_enabled: { $set: value },
                }))}
            />
            {formData.autostart_enabled && (
                    <AutostartHoursForm
                        formData={formData}
                        setFormValue={setFormValue}
                        disabled={disabled}
                    />
            )}
        </>
    );
}
