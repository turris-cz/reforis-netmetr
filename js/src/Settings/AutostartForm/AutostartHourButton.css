/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

.hour-buttons {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    margin-bottom: 1rem;
}

.hour-buttons .btn {
    width: 7rem;
    margin: 0.2rem;
}
