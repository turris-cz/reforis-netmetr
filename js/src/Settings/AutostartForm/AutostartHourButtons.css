/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

.autostart-hours-table td {
    vertical-align: middle;
    horiz-align: center;
}

.btn-group {
    margin-bottom: 1rem;
}
