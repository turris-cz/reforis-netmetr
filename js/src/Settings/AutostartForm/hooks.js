/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { ALL_HOURS } from "./constants";

// Due to poor design of setFormValue function we need to pass there fake event to get it worked.
const fakeEvent = { target: { value: undefined } };

export default function useAutostartHours(hours, setFormValue) {
    function toggleHour(hour) {
        if (hours.includes(hour)) {
            deleteHour(hour);
        } else {
            addHour(hour);
        }
    }

    function addHour(hour) {
        updateHours({ $push: [hour] });
    }

    function deleteHour(hour) {
        const hourIndex = hours.findIndex((currentHour) => currentHour === hour);
        updateHours({ $splice: [[hourIndex, 1]] });
    }

    function selectAllHours() {
        updateHours({ $set: ALL_HOURS });
    }

    function unselectAllHours() {
        updateHours({ $set: [] });
    }

    function updateHours(updateRule) {
        setFormValue(() => (
            { hours_to_run: updateRule }
        ))(fakeEvent);
    }

    return [toggleHour, selectAllHours, unselectAllHours];
}
