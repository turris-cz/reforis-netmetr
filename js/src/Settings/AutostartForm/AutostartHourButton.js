/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { toLocaleDateString } from "foris";
import { useUID } from "react-uid";
import PropTypes from "prop-types";

import "./AutostartHourButton.css";

AutostartHourButton.propTypes = {
    hour_to_run: PropTypes.number.isRequired,
    checked: PropTypes.bool.isRequired,
};

export default function AutostartHourButton({
    hour_to_run,
    checked,
    ...props
}) {
    const uid = useUID();

    const createdAt = toLocaleDateString(hour_to_run, {
        inputFormat: "H",
        outputFormat: "LT",
    });

    return (
        <label
            htmlFor={uid}
            className={`btn btn-outline-secondary ${checked ? "active " : ""}`}
        >
            {createdAt}
            <input id={uid} type="checkbox" autoComplete="off" {...props} />
        </label>
    );
}
