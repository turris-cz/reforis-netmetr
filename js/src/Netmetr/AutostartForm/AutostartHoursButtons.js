/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import "./AutostartHourButtons.css";
import { Button } from "foris";
import AutostartHourButton from "./AutostartHourButton";
import useAutostartHours from "./hooks";
import { ALL_HOURS } from "./constants";

AutostartHoursButtons.propTypes = {
    formData: PropTypes.shape({
        autostart_enabled: PropTypes.bool.isRequired,
        hours_to_run: PropTypes.arrayOf(PropTypes.number).isRequired,
    }).isRequired,
    setFormValue: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

export default function AutostartHoursButtons({ formData, setFormValue, disabled }) {
    const [
        toggleHour,
        selectAllHours,
        unselectAllHours,
    ] = useAutostartHours(formData.hours_to_run, setFormValue);

    return (
        <>
            <div className="btn-group">
                <Button
                    className="btn-sm btn-outline-info "
                    onClick={selectAllHours}
                    disabled={disabled}
                >
                    {_("Select all")}
                </Button>
                <Button
                    className="btn-sm btn-outline-info"
                    onClick={unselectAllHours}
                    disabled={disabled}
                >
                    {_("Unselect all")}
                </Button>
            </div>
            <div className="btn-group-toggle hour-buttons" data-toggle="buttons">
                {ALL_HOURS.map(
                    (hour) => (
                        <AutostartHourButton
                            key={hour}
                            hour_to_run={hour}
                            checked={formData.hours_to_run.includes(hour)}
                            onClick={() => toggleHour(hour)}
                            disabled={disabled}
                        />
                    ),
                )}
            </div>
        </>
    );
}
