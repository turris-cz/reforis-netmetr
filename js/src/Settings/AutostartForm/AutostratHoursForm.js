/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import AutostartHoursButtons from "./AutostartHoursButtons";

AutostartHoursForm.propTypes = {
    formData: PropTypes.shape({
        autostart_enabled: PropTypes.bool.isRequired,
        hours_to_run: PropTypes.arrayOf(PropTypes.number).isRequired,
    }).isRequired,
    setFormValue: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

export default function AutostartHoursForm({
    formData,
    setFormValue,
    disabled,
}) {
    return (
        <>
            <h3>{_("Select autostart hours")}</h3>
            <AutostartHoursButtons
                formData={formData}
                setFormValue={setFormValue}
                disabled={disabled}
            />
        </>
    );
}
