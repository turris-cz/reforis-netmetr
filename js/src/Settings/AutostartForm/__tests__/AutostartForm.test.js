/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {render} from "foris/testUtils/customTestRender";

import AutostartForm from '../AutostartForm';
import autostartFixture from '../../__tests__/__fixtures__/autostart';

describe("<AutostartForm />", () => {
    it("should render component", () => {
        const setFormValue = jest.fn(() => jest.fn());
        const {getByText, container} = render(
            <AutostartForm
                formData={autostartFixture}
                setFormValue={setFormValue}
            />
        );
        expect(getByText("Autostart")).toBeDefined();
        expect(container).toMatchSnapshot();
    });
});
