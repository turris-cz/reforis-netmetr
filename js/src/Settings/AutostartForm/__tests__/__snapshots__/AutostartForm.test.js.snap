// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutostartForm /> should render component 1`] = `
<div>
  <h3>
    Autostart
  </h3>
  <div
    class="form-group"
  >
    <div
      class="custom-control custom-checkbox "
    >
      <input
        checked=""
        class="custom-control-input"
        id="1"
        type="checkbox"
      />
      <label
        class="custom-control-label"
        for="1"
      >
        Enabled
      </label>
    </div>
  </div>
  <h4>
    Select autostart hours
  </h4>
  <div
    class="btn-group"
  >
    <button
      class="btn btn-sm btn-outline-info "
      type="button"
    >
       
       
      Select all
    </button>
    <button
      class="btn btn-sm btn-outline-info"
      type="button"
    >
       
       
      Unselect all
    </button>
  </div>
  <div
    class="btn-group-toggle hour-buttons"
    data-toggle="buttons"
  >
    <label
      class="btn btn-outline-light active "
      for="2"
    >
      12:00 AM
      <input
        autocomplete="off"
        id="2"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light active "
      for="3"
    >
      1:00 AM
      <input
        autocomplete="off"
        id="3"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light active "
      for="4"
    >
      2:00 AM
      <input
        autocomplete="off"
        id="4"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light active "
      for="5"
    >
      3:00 AM
      <input
        autocomplete="off"
        id="5"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light active "
      for="6"
    >
      4:00 AM
      <input
        autocomplete="off"
        id="6"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="7"
    >
      5:00 AM
      <input
        autocomplete="off"
        id="7"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="8"
    >
      6:00 AM
      <input
        autocomplete="off"
        id="8"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="9"
    >
      7:00 AM
      <input
        autocomplete="off"
        id="9"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="10"
    >
      8:00 AM
      <input
        autocomplete="off"
        id="10"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="11"
    >
      9:00 AM
      <input
        autocomplete="off"
        id="11"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="12"
    >
      10:00 AM
      <input
        autocomplete="off"
        id="12"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="13"
    >
      11:00 AM
      <input
        autocomplete="off"
        id="13"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="14"
    >
      12:00 PM
      <input
        autocomplete="off"
        id="14"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="15"
    >
      1:00 PM
      <input
        autocomplete="off"
        id="15"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="16"
    >
      2:00 PM
      <input
        autocomplete="off"
        id="16"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="17"
    >
      3:00 PM
      <input
        autocomplete="off"
        id="17"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="18"
    >
      4:00 PM
      <input
        autocomplete="off"
        id="18"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="19"
    >
      5:00 PM
      <input
        autocomplete="off"
        id="19"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="20"
    >
      6:00 PM
      <input
        autocomplete="off"
        id="20"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="21"
    >
      7:00 PM
      <input
        autocomplete="off"
        id="21"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="22"
    >
      8:00 PM
      <input
        autocomplete="off"
        id="22"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="23"
    >
      9:00 PM
      <input
        autocomplete="off"
        id="23"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="24"
    >
      10:00 PM
      <input
        autocomplete="off"
        id="24"
        type="checkbox"
      />
    </label>
    <label
      class="btn btn-outline-light "
      for="25"
    >
      11:00 PM
      <input
        autocomplete="off"
        id="25"
        type="checkbox"
      />
    </label>
  </div>
</div>
`;
